### ISSUE TITLE
PHP: Sanitize by side effect does not work

### ISSUE THREAD

**Issue by dozernz** [semgrep/semgrep#4277]
**Describe the bug**
I cannot create a PHP rule in taint mode that uses the side-effect sanitizing feature (as per https://r2c.dev/blog/2021/taint-mode-is-now-in-beta/#sanitized-by-side-effect).

**To Reproduce**
https://semgrep.dev/s/dozernz:phpsideeffect

**Expected behavior**
The $vuln var should be matched , but not the $notvuln var. However both are matched.

**What is the priority of the bug to you?**

- [ ] P0: blocking your adoption of Semgrep or workflow
- [X] P1: important to fix or quite annoying
- [ ] P2: regular bug that should get fixed

**Comment by IagoAbal** [comment]
Hi @dozernz, I fixed your example here: https://semgrep.dev/s/Dz1o

Thanks for taking the time to open an issue, I will clarify this in the docs. BTW, I am just curious, are you new to Semgrep and trying taint mode? Please let me know if I can help you with other taint rules, and if you want we can chat via r2c's community Slack. I am very interested in gathering feedback around taint-mode.

Note that a sanitizer is just a Semgrep query that will match some code, and the code that is matched will be annotated as "sanitized".

The sanitizer you wrote was:

```yaml
    pattern-sanitizers:
      - patterns:
          - pattern-inside: sanitize($VAL)
          - pattern: $VAL
```

Note that the `pattern-inside: sanitize($VAL)` only matches expressions that correspond to calls to the function `sanitize`. This query will find expressions like `sanitize(x)` and annotate `x` as sanitized.

So when you have this code:

```php
$notused = sanitize($var);
$notvuln = sink($var);
```

Your sanitizer will only match the first line, where `sanitize` is called, and it will sanitize the first occurrence of `$var`, but not the second!

What the "cookbook" suggests you to do is:

```yaml
    pattern-sanitizers:
      - patterns:
          - pattern-inside: sanitize($VAL); ...
          - pattern: $VAL
```

Note the extra `; ...` in the `pattern-inside`. Now `pattern-inside: sanitize($VAL); ...` matches a sequence of statements that start with a call to `sanitize`. And within that code, it will annotate all occurrences of `$VAL` as sanitized.

So, again, given this code:

```php
$notused = sanitize($var);
$notvuln = sink($var);
```

The  `pattern-inside: sanitize($VAL); ...` matches both lines and `$VAL` is bound to `$var`, and both occurrences of `$var` become sanitized.

I understand this can be subtle especially if you are new to Semgrep. Right now we don't really support sanitization by side-effect and instead we "abuse" Semgrep's `...` for this purpose. I hope this explanation helped!

**Comment by dozernz** [comment]
Hi,

Thanks for clarifying and creating a fixed example, I can see how it works now. When I read the blog post I had incorrectly assumed the `...` on the newline was a placeholder for other rules, rather than actually part of the rule.

> are you new to Semgrep and trying taint mode

Yes, I have only used it once before. This was the first complex rule I tried to write using taint mode (which is very helpful). If I have any more questions I will definitely drop by the Slack.  

Thanks!
